main.go

```diff
@@ -699,160 +528,49 @@ func (c *config) offsetSelection(file ast.Node) (int, int, error) {
 	}
 
 	// offset selects all fields
-	start := c.fset.Position(encStruct.Pos()).Line
-	end := c.fset.Position(encStruct.End()).Line
-
-	return start, end, nil
+	return encStruct.Pos(), encStruct.End(), nil
 }
 
 // allSelection selects all structs inside a file
-func (c *config) allSelection(file ast.Node) (int, int, error) {
+func (cfg *config) allSelection(file *ast.File) (token.Pos, token.Pos, error) {
 	start := 1
-	end := c.fset.File(file.Pos()).LineCount()
+	end := cfg.fset.File(file.Pos()).LineCount()
 
-	return start, end, nil
-}
-
-func isPublicName(name string) bool {
-	for _, c := range name {
-		return unicode.IsUpper(c)
+	// Convert start and end line numbers to token.Pos.
+	tokFile := cfg.fset.File(file.FileStart)
+	lineCount := tokFile.LineCount()
+	if start > lineCount || end > lineCount {
+		return 0, 0, errors.New("line selection is invalid")
 	}
-	return false
-}
 
-// rewrite rewrites the node for structs between the start and end
-// positions
-func (c *config) rewrite(node ast.Node, start, end int) (ast.Node, error) {
-	errs := &rewriteErrors{errs: make([]error, 0)}
-
-	rewriteFunc := func(n ast.Node) bool {
-		x, ok := n.(*ast.StructType)
-		if !ok {
-			return true
-		}
-
-		for _, f := range x.Fields.List {
-			line := c.fset.Position(f.Pos()).Line
-
-			if !(start <= line && line <= end) {
-				continue
-			}
-
-			fieldName := ""
-			if len(f.Names) != 0 {
-				for _, field := range f.Names {
-					if !c.skipUnexportedFields || isPublicName(field.Name) {
-						fieldName = field.Name
-						break
-					}
-				}
-			}
-
-			// anonymous field
-			if f.Names == nil {
-				ident, ok := f.Type.(*ast.Ident)
-				if !ok {
-					continue
-				}
-
-				if !c.skipUnexportedFields {
-					fieldName = ident.Name
-				}
-			}
-
-			// nothing to process, continue with next line
-			if fieldName == "" {
-				continue
-			}
-
-			if f.Tag == nil {
-				f.Tag = &ast.BasicLit{}
-			}
-
-			res, err := c.process(fieldName, f.Tag.Value)
-			if err != nil {
-				errs.Append(fmt.Errorf("%s:%d:%d:%s",
-					c.fset.Position(f.Pos()).Filename,
-					c.fset.Position(f.Pos()).Line,
-					c.fset.Position(f.Pos()).Column,
-					err))
-				continue
-			}
-
-			f.Tag.Value = res
-		}
-
-		return true
-	}
-
-	ast.Inspect(node, rewriteFunc)
-
-	c.start = start
-	c.end = end
-
-	if len(errs.errs) == 0 {
-		return node, nil
-	}
-
-	return node, errs
+	return tokFile.LineStart(start), token.Pos(tokFile.Size()), nil
```

Comment:
This is also incorrect, for the same reason as above: it should be tokFile.Pos(tokFile.Size()).

(again, due to the way this package works, it's just an off-by-one error).